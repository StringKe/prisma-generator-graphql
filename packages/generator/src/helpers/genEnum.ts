import { DMMF } from '@prisma/generator-helper'
import { getComment } from '../utils/getComment'

/**
 * generate graphql enum
 * @param name
 * @param values
 * @param documentation
 */
export const genEnum = ({
  name,
  values,
  documentation,
}: DMMF.DatamodelEnum) => {
  const enumValues = values.map(({ name }) => `\t${name}`).join('\n')

  return `${getComment(documentation)}\nenum ${name} { \n${enumValues}\n }\n`
}
