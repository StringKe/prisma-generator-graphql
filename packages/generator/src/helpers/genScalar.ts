import { DMMF } from '@prisma/generator-helper'
import { getComment } from '../utils/getComment'

/**
 * generate graphql type
 * @param models
 */
export const genScalar = (models: DMMF.Model[]) => {
  const basicSupportType = ['String', 'Int', 'Float', 'Boolean', 'ID']
  const modelNames = models.map((model) => {
    return model.name
  })
  const scalarTypes = models
    .map((model) => {
      return model.fields.map((field) => {
        return field.type
      })
    })
    .flat()
    .filter((type) => {
      return ![...basicSupportType, ...modelNames].includes(type)
    })
    .filter((type, index, self) => {
      return self.indexOf(type) === index
    })
    .map((type) => {
      return `scalar ${type}`
    })
    .join('\n')

  return `\n${scalarTypes}\n`
}
