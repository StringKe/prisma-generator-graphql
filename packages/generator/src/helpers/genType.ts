import { DMMF } from '@prisma/generator-helper'
import { getComment } from '../utils/getComment'

/**
 * generate graphql type
 * @param name
 * @param fields
 * @param documentation
 */
export const genType = ({ name, fields, documentation }: DMMF.Model) => {
  const typeFields = fields
    .map(({ name, type, isRequired, isList, documentation }) => {
      const required = isRequired ? '!' : ''
      const combineType = isList ? `[${type}]` : type
      return `${getComment(documentation)}\t${name}: ${combineType}${required}`
    })
    .join('\n')

  return `${getComment(documentation)}type ${name} { \n${typeFields}\n }\n`
}
