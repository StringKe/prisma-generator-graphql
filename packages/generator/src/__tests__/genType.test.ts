import { getSampleDMMF } from './__fixtures__/getSampleDMMF'
import { genType } from '../helpers/genType'

test('type generation', async () => {
  const sampleDMMF = await getSampleDMMF()

  sampleDMMF.datamodel.models.forEach((model) => {
    expect(genType(model)).toMatchSnapshot(model.name)
  })
})
