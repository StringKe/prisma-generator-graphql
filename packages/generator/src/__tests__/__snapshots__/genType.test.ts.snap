// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type generation: Account 1`] = `
"type Account { 
	id: Int!
	email: String!
	name: String
	user: User
	userId: Int
 }
"
`;

exports[`type generation: User 1`] = `
"type User { 
	id: Int!
	email: String!
	name: String
	test: Json
	Account: [Account]!
 }
"
`;
